### ISSUE TITLE
[legacy][nov20] Installation breaks at second execution

### ISSUE THREAD

**Issue by fuhlig1** [FairRootGroup/FairSoft#370]
When there was a problem with the installation of FairSoft legacy nov20 (e.g. some package couldn't be compiled) the execution of make stops. With each subesquent call of make I run into the problem that packages which were already completely installed in the first run of make without any problems now crashes. An example of the problem is shown below.
Faltbuffers and DDS are already installed but make tries to reinstall them and finally run into the problem that some patches can't be applied because they were already applied in the first run of make. 
In my opinion the underlying problem is that make doesn't realize that the package was alredy installed properly.    

```
cmake --build $PWD/build -j4
[  4%] Built target vc
[  9%] Built target hepmc
[ 13%] Built target pythia6
[ 18%] Built target boost
[ 22%] Built target clhep
[ 22%] Performing update step for 'zeromq'
[ 27%] Built target fmt
[ 27%] Performing update step for 'flatbuffers'
[ 31%] Built target geant4
[ 36%] Built target pythia8
-- flatbuffers update command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/flatbuffers-update-*.log
-- zeromq update command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/zeromq-update-*.log
[ 36%] Performing patch step for 'flatbuffers'
[ 36%] No patch step for 'zeromq'
[ 36%] Performing update step for 'dds'
[ 36%] Performing update step for 'fairlogger'
CMake Error at /opt/fairsoft/source/nov20/build/Stamp/flatbuffers/flatbuffers-patch-RelWithDebInfo.cmake:37 (message):
  Command failed: 1

   '/usr/bin/patch' '-N' '-p1' '-i' '/opt/fairsoft/source/nov20/legacy/flatbuffers/remove_werror.patch'

  See also

    /opt/fairsoft/source/nov20/build/Log/flatbuffers-patch.log


-- Log output is:
patching file CMakeLists.txt
Reversed (or previously applied) patch detected!  Skipping patch.
3 out of 3 hunks ignored -- saving rejects to file CMakeLists.txt.rej

CMake Error at /opt/fairsoft/source/nov20/build/Stamp/flatbuffers/flatbuffers-patch-RelWithDebInfo.cmake:47 (message):
  Stopping after outputting logs.


make[2]: *** [Stamp/flatbuffers/flatbuffers-patch] Error 1
make[1]: *** [CMakeFiles/flatbuffers.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 40%] Performing configure step for 'zeromq'
-- fairlogger update command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/fairlogger-update-*.log
-- dds update command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/dds-update-*.log
[ 40%] No patch step for 'fairlogger'
[ 40%] Performing patch step for 'dds'
CMake Error at /opt/fairsoft/source/nov20/build/Stamp/dds/dds-patch-RelWithDebInfo.cmake:37 (message):
  Command failed: 1

   '/usr/bin/patch' '-N' '-p1' '-i' '/opt/fairsoft/source/nov20/legacy/dds/fix_boost_lookup.patch'

  See also

    /opt/fairsoft/source/nov20/build/Log/dds-patch.log


-- Log output is:
patching file CMakeLists.txt
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file CMakeLists.txt.rej

CMake Error at /opt/fairsoft/source/nov20/build/Stamp/dds/dds-patch-RelWithDebInfo.cmake:47 (message):
  Stopping after outputting logs.


make[2]: *** [Stamp/dds/dds-patch] Error 1
make[1]: *** [CMakeFiles/dds.dir/all] Error 2
[ 40%] Performing configure step for 'fairlogger'
-- fairlogger configure command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/fairlogger-configure-*.log
[ 45%] Performing build step for 'fairlogger'
-- fairlogger build command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/fairlogger-build-*.log
[ 45%] Performing install step for 'fairlogger'
-- fairlogger install command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/fairlogger-install-*.log
[ 45%] Completed 'fairlogger'
[ 45%] Built target fairlogger
-- zeromq configure command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/zeromq-configure-*.log
[ 45%] Performing build step for 'zeromq'
-- zeromq build command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/zeromq-build-*.log
[ 45%] Performing install step for 'zeromq'
-- zeromq install command succeeded.  See also /opt/fairsoft/source/nov20/build/Log/zeromq-install-*.log
[ 45%] Completed 'zeromq'
[ 45%] Built target zeromq

```

**Comment by dennisklein** [comment]
See https://github.com/FairRootGroup/FairSoft/blob/7e1412f4335f8b86dd2b80d6a484da251d3ed01f/cmake/legacy.cmake#L17.

Does your `patch` command understand the `-N` option? I think I added it exactly because of the behaviour you describe here.

**Comment by fuhlig1** [comment]
Yes my patch command support the flag, at least 'patch --help' shows the option. When I apply the patch using the command the patch is skipped but I nevertheless get a return value of 1 which for cmake indicates that the command fails.

> '/usr/bin/patch' '-N' '-p1' '-i' '/opt/fairsoft/source/nov20/legacy/dds/fix_boost_lookup.patch'
> patching file CMakeLists.txt
> Reversed (or previously applied) patch detected!  Skipping patch.
> 1 out of 1 hunk ignored -- saving rejects to file CMakeLists.txt.rej
> demac006:dds uhlig$ echo $?
> 1
 
Why it is needed to run through all possible stages when the install stage was successful in a previous run?

**Comment by dennisklein** [comment]
I just re-checked on mac and linux, the `-N` seems to not work as expected :( However, the `--merge` option of GNU patch seems to do the right thing (skipping if already applied).

> Why it is needed to run through all possible stages when the install stage was successful in a previous run?

That is how the authors of `ExternalProject_Add` have decided to implement the patch step. It requires an idempotent patch command. I will try to come up with a more robust command.

**Comment by dennisklein** [comment]
* GNU `patch --merge` seems to skip already applied hunks and returns success (zero)

```
$ patch --merge -p1 -i patch blubb; echo $?
patching file blubb
Hunk #1 already applied at 2.
0
```

* GNU and BSD `patch -N -r-` seem to skip already applied hunks and the generation of a `.rej` file but exit with non-zero (one)

```
$ patch -N -r- -p1 -i patch blubb; echo $?
patching file blubb
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
1
```

**Comment by fuhlig1** [comment]
A workaround is to install the brew package "gpatch" which installs the GNU version of patch and use this one together with the option "--merge" in cmake/legacy.cmake.

**Comment by dennisklein** [comment]
> A workaround is to install the brew package "gpatch" which installs the GNU version of patch and use this one together with the option "--merge" in cmake/legacy.cmake.

I will adopt this solution, thx!

**Comment by fuhlig1** [comment]
How do you want to do that?

I would propose to add the package to the list of dependencies. In this case you can simply change the patch option "-N" to "--merge".

**Comment by dennisklein** [comment]
The fix is included in the `nov20_patches` branch (f14e59b87460a2b1954f0a6008feee6937c4ac56) and will be included in the next patch release.
